initial commit (#5)
aswinayyolath/stretch-cluster-testing: testing/performance-testing/performance-testing.md

# Performance Testing

## Open messaging benchmark framework

> The Open messaging benchmark framework was used for benchmarking the stretch kafka clusters. By following this tutorial https://community.ibm.com/community/user/integration/blogs/dale-lane1/2025/02/03/running-openmessaging-benchmarks we were able to get benchmark results for three configurations

Reviewer: ```suggestion
The Open Messaging Benchmark framework was used to benchmark the stretch Kafka clusters. By following this [tutorial](https://community.ibm.com/community/user/integration/blogs/dale-lane1/2025/02/03/running-openmessaging-benchmarks), we obtained benchmark results for three configurations:
```

>  - Regular kafka cluster

Reviewer: ```suggestion
 - Regular Kafka cluster
```

>  - Stretch kafka cluster using submariner

Reviewer: ```suggestion
 - Stretch kafka cluster with submariner
```

>  - Stretch kafka cluster using submariner

Reviewer: ```suggestion
 - Stretched Kafka cluster using Submariner
```

>  - Stretch kafka cluster using cilium

Reviewer: ```suggestion
 - Stretch kafka cluster with cilium
```

>  - Stretch kafka cluster using cilium

Reviewer: ```suggestion
 - Stretched Kafka cluster using Cilium
```

> The Open messaging benchmark framework deploys 8 benchmark workers and one driver to the K8s environment. Once the tests start, the driver employs the workers to produce/consume messages to the cluster using its bootstrap url.

Reviewer: ```suggestion
The Open messaging benchmark framework deploys 8 benchmark workers and one driver in the K8s environment. Once the tests start, the driver employs the workers to produce/consume messages to the cluster using its bootstrap url.
```

> The Open messaging benchmark framework deploys 8 benchmark workers and one driver to the K8s environment. Once the tests start, the driver employs the workers to produce/consume messages to the cluster using its bootstrap url.

Reviewer: ```suggestion
The Open Messaging Benchmark framework deploys eight benchmark workers and one driver to the Kubernetes (K8s) environment. Once the tests start, the driver employs the workers to produce and consume messages in the cluster using its bootstrap URL.
```

## Configuration

> For all the stretched kafka cluster configurations, the OCP/K8s clusters were in the same data centers hence minimizing the network latency and shifting the focus on the additional latency provided by the tools in use.

Reviewer: ```suggestion
For all the stretched Kafka cluster configurations, the OpenShift (OCP)/K8s clusters were located in the same data centers, minimizing network latency and focusing on the additional latency introduced by the networking tools in use.
```

### Regular Kafka cluster

> We have used a regular kraft kafka cluster using 3 brokers and 3 controllers and ephemeral storage. This CR was applied in an OCP environment.

Reviewer: ```suggestion
We used a regular Kraft Kafka cluster with three brokers, three controllers, and ephemeral storage. This Custom Resource (CR) was applied in an OCP environment.
```

  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - controller
  storage:
    type: jbod
    volumes:
      - id: 0
        type: ephemeral
        kraftMetadata: shared
---

apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  name: broker
  labels:
    strimzi.io/cluster: my-cluster
spec:
  replicas: 3
  roles:
    - broker
  storage:
    type: jbod
    volumes:
      - id: 0
        type: ephemeral
        kraftMetadata: shared
---

apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/node-pools: enabled
    strimzi.io/kraft: enabled
spec:
  kafka:
    version: 3.8.0
    metadataVersion: 3.8-IV0
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
  entityOperator:
    topicOperator: {}
    userOperator: {}
```

### Stretched kafka cluster with submariner

> The kafka cluster was deployed in three OCP environment. We havd 3 brokers and 3 controllers in all three OCP clusters. Similar to the regular stretch cluster, we've used ephemeral storage for this configuration. Submariner was deployed on all three clusters and connected using the `subctl` binary.

Reviewer: ```suggestion
The Kafka cluster was deployed across three OCP environments, each hosting three brokers and three controllers. As with the regular Kafka cluster, we used ephemeral storage. Submariner was deployed on all three clusters and connected using the `subctl` binary.
```

metadata:
  name: stretch2-broker
  labels:
    strimzi.io/cluster: my-cluster
    strimzi.io/submariner-cluster-id: "cluster3"
spec:
  cluster: cluster-b #-------> all resources part of this KNP will be created in K8s cluster-b
  replicas: 3
  roles:
    - broker
  storage:
    type: jbod
    volumes:
      - id: 0
        type: ephemeral
        kraftMetadata: shared
---

apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  name: my-cluster
  annotations:
    strimzi.io/node-pools: enabled
    strimzi.io/kraft: enabled
    strimzi.io/cross-cluster-type: "submariner"  #------->  Cross Cluster Communication Technology in use 
spec:
  kafka:
    version: 3.8.0
    metadataVersion: 3.8-IV0
    listeners:
      - name: plain
        port: 9092
        type: internal
        tls: false
      - name: tls
        port: 9093
        type: internal
        tls: true
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
  entityOperator:
    topicOperator: {}
    userOperator: {}
```

### Stretched kafka cluster with Cilium

> Three k8s clusters were connected using the `cilium` cli and then stretch was implemented on this using the exact same CR used for submariner. 3 brokers, 3 controllers on each of the clusters working together to form a stretched kafka cluster using cilium.

Reviewer: ```suggestion
Three K8s clusters were connected using the `cilium` cli and then stretch was implemented on this using the exact same CR used for submariner. 3 brokers, 3 controllers on each of the clusters working together to form a stretched kafka cluster using cilium.
```

## Findings

> Three iterations of tests were taken for each environment to avoid any environement related noise from the test results. These results were taken to a json file and plotted using matplotlib. 4 parameters were of signifance to us.

Reviewer: ```suggestion
Three test iterations were performed for each environment to eliminate environmental noise. The results were exported as JSON files and plotted using Matplotlib. The following four parameters were analyzed:
```

1. Publish Delay (ms)
2. End to end Delay (ms)
3. Publish Rate (messages/sec)
4. Consume Rate (messages/sec)

> These results are copied to folders non-stretch, stretch-cilium and stretch-submariner.

Reviewer: ```suggestion
The results were stored in directories named non-stretch, stretch-cilium, and stretch-submariner.
```

> The plot for these look something like this:

Reviewer: ```suggestion
The plotted results are shown below:
```

![performance-plot](performance-plot.png)

> The three plots for each color represents the plot for 3 iterations for each configurations.

Reviewer: ```suggestion
Each color in the graph represents the three test iterations for each configuration
```

> The findings are clear. The stretched kafka cluster with submariner provided too much additional latency to be comparable with the regular kafka cluster while cilium provides comparable levels of performance with the regular kafka cluster.

Reviewer: ```suggestion

## Conclusion
The results indicate that:

- The stretched Kafka cluster with Submariner introduces significant additional latency, making it less comparable to the regular Kafka cluster.
- The stretched Kafka cluster with Cilium provides performance levels comparable to the regular Kafka cluster.

Thus, Cilium appears to be the better choice for implementing stretched Kafka clusters while maintaining acceptable performance levels.
```
